### ISSUE TITLE
During indexer:reindex catalogsearch_fulltext no products are listed in categories

### ISSUE THREAD

**Issue by heldchen** [magento/magento2#6754]
while the "Catalog Search" is being reindexed, all category listing pages show "We can't find products matching the selection.". as soon as the reindex process is completed, the categories are browsable again.
### Preconditions
1. CE 2.1.1
2. full page cache is disabled
### Steps to reproduce
1. create a large sized sample database with enough products that the fulltext reindex takes a while. ours has around 20'000 products with lots of attributes and takes 4min17s on a beefy server to reindex
2. `php bin/magento indexer:reindex catalogsearch_fulltext`
3. meanwhile, browse the fronted categories
### Expected result
1. the categories list products during reindexing
### Actual result
1. all categories show a "We can't find products matching the selection." message

as mentioned in #6308 this is a regression compared to the magento 1 behaviour. magento 2 cannot assume during a categories reindex that a full page cached version of the view exists and thus no error would be shown to the user; this error is also shown with full page cache enabled but then browsing a non-cached url.

the fulltext index is just one of the indexes that will show this behavior, but the easiest to reproduce it with as the others are much faster reindexed.

**Comment by stamster** [comment]
What happens if you previously enable FPC?

**Comment by heldchen** [comment]
from my tests, enabling the FPC followed by an immediate reindex exhibits the same problems and makes no difference - IMO this is expected as the FPC at this point is not primed.

even if that worked, FPC never was a requirement for a working setup so far (i.e. in magento 1.x). it's not like we do not want an FPC, but enabling the internal magento 2 full page cache in production mode is advised against by the devdocs, while open issues like wrongly cached breadcrumbs f.e. prevent us from using a proper varnish cache in production.

**Comment by sevos1984** [comment]
This is a known issue. Internal ticket MAGETWO-58174. Thanks for reporting.

**Comment by heldchen** [comment]
thanks!

FWIW: this also happens with FPC enabled. 

with FPC enabled it's even worse as it might happen that the "We can't find products matching the selection."-page is cached and served, even after the reindex finished!

**Comment by peexl-dev** [comment]
also happens in 2.1.3 version

**Comment by hostep** [comment]
We also run into this, in case anyone is interested, here are all the commits tagged as MAGETWO-58174:

- https://github.com/magento/magento2/commit/767944cd522
- https://github.com/magento/magento2/commit/4638e6febcd
- https://github.com/magento/magento2/commit/dbbd83dad7e
- https://github.com/magento/magento2/commit/630c34596dd
- https://github.com/magento/magento2/commit/a0538725bd7
- https://github.com/magento/magento2/commit/17dffbe3b95
- https://github.com/magento/magento2/commit/b04ff64fc28
- https://github.com/magento/magento2/commit/8b967272fb3
- https://github.com/magento/magento2/commit/cc04f5abb53
- https://github.com/magento/magento2/commit/a29406dc4a0
- https://github.com/magento/magento2/commit/2ee18cd557d
- https://github.com/magento/magento2/commit/e4f0890d823
- https://github.com/magento/magento2/commit/9784cfc59ad

I think the last commit is a combination of all the above, but not 100% sure.

I haven't had the time yet to see if this is backportable to Magento 2.1.x, maybe I'll try this out later...

**Comment by eldonbite** [comment]
Hi @hostep,

I tried creating a patch file (.diff format) using the commits for the said internal ticket. I applied it to one of our projects and we are currently testing it. In case you're interested, I'm attaching it here. Anyway I also need some help in verifying whether these commits can really fix this issue.

[GIT-6754.diff.txt](https://github.com/magento/magento2/files/812987/GIT-6754.diff.txt)

We're using Magento 2.1.3 by the way.

**Comment by hostep** [comment]
Hi @eldonbite 

Sorry for the late answer.
We also created a diff which we can apply to Magento 2.1.2 (also tested on 2.1.3 and 2.1.5)

I just took a look at the difference between your diff and ours, and it's almost exactly the same. There are some minor differences in the Tests (we don't run the tests, so there might be something wrong in our diff). But there is one big difference in the `etc/di.xml` file.
In your patch, you are adding:
```diff
+    <preference for="Magento\CatalogSearch\Model\Search\FilterMapper\FilterStrategyInterface" type="Magento\CatalogSearch\Model\Search\FilterMapper\FilterContext"/>
```
and
```diff
+    <type name="Magento\CatalogSearch\Model\Search\RequestGenerator\GeneratorResolver">
+        <arguments>
+            <argument name="defaultGenerator" xsi:type="object">\Magento\CatalogSearch\Model\Search\RequestGenerator\General</argument>
+            <argument name="generators" xsi:type="array">
+                <item name="decimal" xsi:type="object">Magento\CatalogSearch\Model\Search\RequestGenerator\Decimal</item>
+            </argument>
+        </arguments>
+    </type>
```

Which we don't do, and I don't really see those lines in the commits above, so I think this is incorrect?

In case anyone is interested, here is our diff:
[MAGETWO-58174.diff.txt](https://github.com/magento/magento2/files/836425/MAGETWO-58174.diff.txt)

We are running this on a Magento 2.1.2 in production for a couple of days now, and it looks like this works and doesn't have negative side effects.

**Comment by eldonbite** [comment]
Hi @hostep. Thanks for the update! Will check it out as well.

**Comment by cizgidekiadam** [comment]
@hostep Not working for me with 2.1.2
There is not an error. It is creating a `catalogsearch_fulltext_scope1_tmp` but not filling it.
Another issue, there is fifty thousands records in `catalogsearch_fulltext_cl`, indexer not processing them.

**Comment by zhulak** [comment]
Hi, I'm facing the same issue, no products visible in front end during re-indexing. We are using CE 2.1.7. Any idea in which release the fix will be included?

**Comment by cizgidekiadam** [comment]
Same in 2.1.8 :(

**Comment by josefbehr** [comment]
Also still experiencing the issue on 2.1.8, a fix for the 2.1 branch would be greatly appreciated.

**Comment by hostep** [comment]
I just had to upgrade our patch which we applied to Magento 2.1.7 to make it compatible with Magento 2.1.9, so in case anyone is interested:
[MAGETWO-58174-for-Magento-2.1.9.patch.txt](https://github.com/magento/magento2/files/1351822/MAGETWO-58174-for-Magento-2.1.9.patch.txt)

I haven't actually tested it yet at this moment, so if anyone wants to try this, be my guest.
I might create a backport for this in a couple of days, but no promises yet.

**Comment by magento-engcom-team** [comment]
@heldchen, thank you for your report.
The issue is already fixed in develop branch, 2.2.0

**Comment by heldchen** [comment]
I'm aware of that. but 2.1 is not 2.2, so why is this closed now?

**Comment by jarhody** [comment]
This seems like a pretty big issue if products disappear from an eCommerce site. Is the understanding correct that every major point release is going to continue to be supported/updated for bug fixes of this nature?

Meaning this should still be fixed in 2.1.10 or later.

**Comment by EvanRijn** [comment]
@hostep Hi! Do you have the patch also for 2.1.8 and 2.1.10 ?

Would be great ;)

**Comment by hostep** [comment]
@EvanRijn, feel free to try the patch for 2.1.9 on 2.1.10, it "might" work, but I haven't tested it yet, so no promises.

**Comment by samuelecarpene** [comment]
@hostep i updated version 2.19 to 2.1.10... The bug is still present. How to fix this bug in version 2.1.10?

**Comment by versdivers** [comment]
Just experienced this issue in 2.2.1 version

**Comment by kristjan-h** [comment]
Thanks for the patch @hostep , seems to fix the problem for us as well.
Just for the record - same issue also reported on #10591 

Just in case, I've attached a version of the patch for composer based installations (tests not included)
[MAGETWO-58174-for-Magento-2.1.9.patch.txt](https://github.com/magento/magento2/files/1600374/MAGETWO-58174-for-Magento-2.1.9.patch.txt)

**Comment by pmathbliss** [comment]
@OnTapKristjan That patch broken the catalogsearch_fulltext reindex.
Here is the reindex after applying the patch. Did you miss a file?

> php bin/magento indexer:reindex catalogsearch_fulltext
PHP Fatal error:  Uncaught Error: Cannot instantiate interface Magento\Framework\Search\Request\IndexScopeResolverInterface in /chroot/home//html/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php:93
Stack trace:
#0 /chroot/home//html/vendor/magento/framework/ObjectManager/Factory/Compiled.php(88): Magento\Framework\ObjectManager\Factory\AbstractFactory->createObject('Magento\\Framewo...', Array)
#1 /chroot/home//html/vendor/magento/framework/ObjectManager/Factory/Compiled.php(130): Magento\Framework\ObjectManager\Factory\Compiled->create('Magento\\Framewo...')
#2 /chroot/home//html/vendor/magento/framework/ObjectManager/Factory/Compiled.php(67): Magento\Framework\ObjectManager\Factory\Compiled->get('Magento\\Framewo...')
#3 /chroot/home//html/vendor/magento/framework/ObjectManager/ObjectManager.php(57): Magento\Framework\ObjectManager\Factory\Compiled->create('Magento\\Cat in /chroot/home//html/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php on line 93

**Comment by id-us** [comment]
hi @hostep,
We’re experiencing the same issue in 2.1.7 :/
Can you share your patch for that version?
10x!

**Comment by hostep** [comment]
@orhabas: should be the one in my [comment from 12 Mar 2017](https://github.com/magento/magento2/issues/6754#issuecomment-285934482), but not entirely sure anymore, it's been too long.

Maybe it becomes time to propose this fix in a PR, so the fix gets included in a stable version of 2.1.x (it already should be included in 2.2.x) instead of keep working with these patches. If someone feels like it, please do so, I probably won't have time in the upcoming days, maybe sometime later though...

**Comment by PetraSwed** [comment]
Having this issue on Magento 2.2.0. When running indexer:reindex as soon as the indexer reaches catalogsearch_fulltext products dissapear from frontend. As the indexer progress the products start to appear again - but in our case that takes about 2 hours!

**Comment by versdivers** [comment]
@PetraSwed  2 hours to reindex? We have customers with shops with over 10.000 products and the reindexing does not take longer then 10 minutes. I think you need to look at your setup and the Magento 2 requirements

**Comment by bvboas** [comment]
@versdivers 10 mins to reindex? We have a customer with one product and one store view and takes less than 10s to reindex. I think you need to look at your setup and the Magento 2 requirements

**Comment by PetraSwed** [comment]
15 000 products, 6 stores. It seem both problems (long reindex time and product disappearing) is connected to implementation of Elastic search (third party module). Using mysql for search - reindex takes 6 minutes and products are left in the category - but search is very slow

**Comment by ssmadsen** [comment]
@PetraSwed 
What elasticsearch module are you using?

**Comment by ssmadsen** [comment]
Facing same issue in Magento 2.1.8 using Mirasvit Elastic Search. 
85000 products, index time: 70minutes. 

Two stores. 

No products while reindexing catalogsearch_fulltext.

Can any confirm the above mentioned fixes actually work?

**Comment by PetraSwed** [comment]
@ssmadsen also Mirasvit Elastic Search. This is what they say:
"Yes, with elastic engine, we always clear indexed data before start full reindex (only full reindex clear data).  
it's behavior/limitation of our implementation."

We avoid doing full reindex of catalogsearch_fulltext during the day. But since we are updating product info, stock and prices from PIM and ERP we need to do full reindex at some point for the changes to show. So I set up cron job to do reindex of all indexes separatelly every 15 minutes except the catalogsearch_fulltext. That one I run via cron only once at night 5 am - when hopefully no visitors are shopping. Not really sure if this is the right approach but I did not find better solution. This is how crontab looks like (I also set high memory limit and execution time because it is critical that the index process goes through):
# Runs 05:00 every day (takes about 2 hours OBS products dissapear from frontend)
0 5 * * * php -dmemory_limit=2G -d max_execution_time=20000 bin/magento indexer:reindex catalogsearch_fulltext

# Runs every 15 minutes
*/15 8-23 * * * php -dmemory_limit=2G -d max_execution_time=10000 bin/magento indexer:reindex design_config_grid
*/15 8-23 * * * php -dmemory_limit=2G -d max_execution_time=10000 bin/magento indexer:reindex customer_grid
*/15 8-23 * * * php -dmemory_limit=2G -d max_execution_time=10000 bin/magento indexer:reindex catalog_category_product
*/15 8-23 * * * php -dmemory_limit=2G -d max_execution_time=10000 bin/magento indexer:reindex catalog_product_category
*/15 8-23 * * * php -dmemory_limit=2G -d max_execution_time=10000 bin/magento indexer:reindex catalog_product_price
*/15 8-23 * * * php -dmemory_limit=2G -d max_execution_time=10000 bin/magento indexer:reindex catalog_product_attribute
*/15 8-23 * * * php -dmemory_limit=2G -d max_execution_time=10000 bin/magento indexer:reindex catalogrule_rule
*/15 8-23 * * * php -dmemory_limit=2G -d max_execution_time=10000 bin/magento indexer:reindex catalogrule_product
*/15 8-23 * * * php -dmemory_limit=2G -d max_execution_time=10000 bin/magento indexer:reindex cataloginventory_stock
*/15 8-23 * * * php -dmemory_limit=2G -d max_execution_time=10000 bin/magento indexer:reindex mst_misspell
As I said not sure if this is perfect but seems to be working for us so far.

**Comment by bvboas** [comment]
@PetraSwed  we have 10k products and 6 store view, we use the wyomind elasticsearch plugin and takes less than 3 mins to fully reindex...

**Comment by ssmadsen** [comment]
@PetraSwed 
Yeah, this solution will probably work as a 'quickfix', but in general, clearing all data from elasticserver before the indexer is completely done and ready to be deployed, is just stupid imo.

Though, we still have issues with disapperaed products when only reindexing catalogsearch_fulltext. But maybe thats whats this post is actually about - and not the issue related to Mirasvit Extension.

But in general, the cron-jobs would be the only work-around for now in this case.

(I've talked to Mirasvit support around 34500 times, but none of them gave me that explanation that you got - but maybe they're were 'afraid' to admit it, who knows)

@bvboas 
I've tried the Elastic extension from Wyomind as well, actually this has a 'safe reindex' setting, that only deploys the new indexed data, when the indexer is fully ran. 

And i also see a signitifcant index time compared to Mirasvit as you mention:
Wyomind: Around 10-12 minutes.
Mirasvit: 60-70 minutes.

But since Wyomind is not compatible with Amasty Imporved Layered Navigation, this is not a solution for us. Mirasvit supports this extension 100% (So here's an advice to save some time debugging). 

Bugs with Amasty Layered Navigation and Wyomind where several, but some of them were:
1. Only attributes that was set to "searchable" were indexed - and before the filtration would work in category list, all attributes needed to be set to "searchable". If the attribute was not set - it returned "no products found"

2. The search result page - just showed us random products and were useless.

And for Magento. Please just make elasticserver default in your CE version as with EE. 
Mysql-searches just make everyone tired.
